File: front/web/lib/auth-gate.ts
/**
 * auth-gate.ts — the deterministic TS twin of back/runtime/gateway/authgate.go (S61). It
 * composes the authentication gate (lib/authn, layer 0) UPSTREAM of the gateway router
 * (lib/gateway, layer 1). The ordering IS the defense:
 *
 *   authentication (layer 0) → scope (layer 1, projectWall) → zone (layer 1, registry)
 *                                                          ↘ RLS (layer 2, Postgres)
 *
 * An UNAUTHENTICATED call is refused BEFORE the router runs — it reaches no project data
 * (the property done-criterion). The verified identity is propagated into the scope (the
 * request body never sets it). DETERMINISM-FIRST: pure, same input → same decision
 * (pinned by lib/auth-gate.test.ts). The Go package is authoritative; this twin matches it.
 */

import {
	type AuthBlockReason,
	authenticate,
	CODE_UNAUTHENTICATED,
	type Principal,
} from "./authn";
import {
	type GatewayBlockReason,
	type Outcome as GatewayOutcome,
	lookup,
	route,
	type Tool,
} from "./gateway";
import type { Scope, Target } from "./projectWall";

/** The composed outcome — the gateway outcomes plus the upstream UNAUTHENTICATED refusal. */
export type GatedOutcome = GatewayOutcome | "unauthenticated";

/** The composed refusal shape — either an auth refusal (UNAUTHENTICATED) or a gateway one. */
export type GatedBlockReason = AuthBlockReason | GatewayBlockReason;

export interface GatedDecision {
	outcome: GatedOutcome;
	tool?: Tool;
	blockReason?: GatedBlockReason;
}

/**
 * authenticatedRoute mirrors Go Registry.AuthenticatedRoute exactly. It authenticates the
 * principal (refusing UNAUTHENTICATED upstream of routing — reaching no data), then routes
 * under the VERIFIED identity (never the caller-supplied scope identity). Pure + total.
 */
export function authenticatedRoute(
	principal: Principal,
	toolName: string,
	scope: Scope,
	target: Target,
): GatedDecision {
	const t = lookup(toolName);
	const disp =
		t && t.disposition === "truth_write" ? "truth_write" : "below_line";
	const auth = authenticate(principal, disp);
	if (auth.outcome === "unauthenticated") {
		return { outcome: "unauthenticated", blockReason: auth.blockReason };
	}
	// Propagate the VERIFIED identity into the scope — the request body never sets it.
	const verifiedScope: Scope = { ...scope, identity: principal.identity };
	const d = route(verifiedScope, toolName, target);
	return { outcome: d.outcome, tool: d.tool, blockReason: d.blockReason };
}

/**
 * refuseUnauthenticated EXECUTES an unauthenticated call to a truth-write endpoint through
 * the SAME gate the live server applies. It returns the real UNAUTHENTICATED BlockReason
 * (code + severity + explanation + how_to_fix), proving an anonymous call to a
 * truth-write reaches NO data (the property done-criterion). No truth written.
 */
export function refuseUnauthenticated(
	activeProject: string,
	tool: string,
): GatedBlockReason | null {
	const anon: Principal = { identity: "", email: "", provider: "" };
	const d = authenticatedRoute(
		anon,
		tool,
		{ identity: "", activeProject },
		{ projectId: activeProject, claimedIdentity: "" },
	);
	if (d.outcome !== "unauthenticated" || !d.blockReason) return null;
	return d.blockReason;
}

export { CODE_UNAUTHENTICATED };

File: back/runtime/gateway/authgate.go
// authgate.go — the S61 AUTHENTICATION GATE wired UPSTREAM of the S58 router. It is the
// composition seam where the auth core (back/runtime/authn, layer 0) runs BEFORE the
// project-scope/truth-zone walls (Route). The ordering IS the defense:
//
//	authentication (layer 0) → scope (layer 1, projectwall) → zone (layer 1, registry)
//	                                                        ↘ RLS (layer 2, Postgres)
//
// An UNAUTHENTICATED call is refused before Route runs — it reaches NO project data (the
// property done-criterion). Auth is NEVER gateway-only: the resolved principal's identity
// is the SAME value SET LOCAL app.identity gives the RLS (authn.Principal.GUCs), so a
// gateway bug cannot widen what the RLS sees (the two layers redden independently).
//
// gateway imports authn (a leaf); authn never imports gateway (no cycle). DETERMINISM:
// AuthenticatedRoute is pure — same (principal, call) ⇒ same decision.
package gateway

import "github.com/steph-frtech/aidos/back/runtime/authn"

// CodeUnauthenticated re-exports authn's code so the gateway's block surface is one set
// (the panel/SDK enumerate gateway codes; UNAUTHENTICATED is one of them now).
const CodeUnauthenticated = BlockCode(authn.CodeUnauthenticated)

// AuthenticatedRoute runs the authentication gate, then the router. The order is the wall:
//
//  1. UNAUTHENTICATED — an anonymous principal is refused with UNAUTHENTICATED, BEFORE
//     the tool is even looked up. The decision's Tool is nil and no data is touched.
//  2. otherwise the resolved principal's identity is propagated into the call's scope
//     (overriding any client-asserted identity — the gateway trusts the VERIFIED session,
//     never the request body), and the call proceeds to Route (scope then zone).
//
// The principal carries the verified session/JWT subject (Auth.js front → verified by the
// gateway). PURE: no clock, rng, I/O or LLM. Same input ⇒ same RouteDecision.
func (r *Registry) AuthenticatedRoute(p authn.Principal, call Call) RouteDecision {
	// The disposition of the targeted tool decides nothing about authentication (an
	// anonymous caller is refused for any tool); we resolve it only to keep the gate's
	// signature symmetric with the authn core. An unknown tool still authenticates first.
	disp := authn.DispositionBelowLine
	if t, ok := r.Lookup(call.Tool); ok && t.Disposition == DispositionTruthWrite {
		disp = authn.DispositionTruthWrite
	}
	if d := authn.Authenticate(p, disp); d.Outcome == authn.OutcomeUnauthenticated {
		return RouteDecision{
			Outcome:     OutcomeUnauthenticated,
			BlockReason: fromAuthn(d.BlockReason),
		}
	}
	// Propagate the VERIFIED identity into the scope — the request body never sets it.
	call.Scope.Identity = p.Identity
	return r.Route(call)
}

// OutcomeUnauthenticated is the gate's terminal refusal — an anonymous call, refused
// upstream of routing (reaches no data).
const OutcomeUnauthenticated Outcome = "unauthenticated"

// fromAuthn adapts authn's BlockReason into the gateway shape (the code is identical;
// this only widens the type). Total.
func fromAuthn(b *authn.BlockReason) *BlockReason {
	if b == nil {
		return nil
	}
	return &BlockReason{
		Code:        BlockCode(b.Code),
		Severity:    b.Severity,
		Explanation: b.Explanation,
		HowToFix:    b.HowToFix,
	}
}
